Dogstatsd support

bradlangel opened the pull request:
What do you think about this initial design?

bradlangel committed:
Factors out protocol parsing into Base module

Allows us to have a high-level send/io interface without also containing
the protocol parsing functions contained in Metric, Event, and Service
check.

kevinqiu commented on dogstatsd.mli:
```diff
@@ -0,0 +1,140 @@
+  end
+end
+
+(* TODO: replace IO with {Async, Lwt, Sync} *)
```
We should do the fancy JS thing where they define the common interface with an abstract type(is that the right term?) and then just do an include with the correct type (in this case I think it'd be like `Async.t ()`, `Lwt.t ()`, or just `()`). See `Monad.S` in `option.mli`. We should do this in `statsd_client.mli` also I think. It's not a huge deal since there's just 2 versions but I think it's nice to not have to repeat ourselves.

bradlangel replied:
This is the intention of this comment. I've done something similar in the separate packages pull request. I'm still figuring out how it will all work together wrt this extended interface.

kevinqiu commented on dogstatsd.mli:
```diff
@@ -0,0 +1,140 @@
+    type t =
+      { metric_name : string
+      ; metric : typ
+      ; sample_rate : float option
```
The documentation says:
> Sample rates only work with counter, histogram, and timer metrics

Should we encode the sample rate into the typ perhaps? Too heavy handed? Just make a note in the interface and leave the interface it as is?

bradlangel replied:
>Just make a note in the interface and leave the interface it as is?
Going to do this.

kevinqiu commented on dogstatsd.mli:
```diff
@@ -0,0 +1,140 @@
+      : ?tags:Base.Tag.t list
+      -> ?sample_rate:float
+      -> Base.Metric.typ
+      -> string list
```
Why can this take multiple names but `ServiceChek.t_send` can't?

bradlangel replied:
I added this for convenience since it seems like you might send multiple metrics at the same time but I can remove it and we can revisit if it proves to be useful for the user. And I didn't include it for ServiceCheck since I couldn't think of a time where multiple service checks would need to be sent with the same message.

kevinqiu replied:
ah, makes sense

bradlangel committed:
Adds comments to the dogstatsd interface

bradlangel committed:
Modifies Metric.send interface to send one metric

bradlangel committed:
Move dogstatsd.mli to statsd-client-extended dir.

bradlangel committed:
Adds unimplemented dogstatsd.ml

bradlangel committed:
Adds barebones jbuild file and dogstatsd.opam file

brendanlong reviewed (approved):
I think we should probably land this and try using it, and we'll improve the interface as we actually use it.

bradlangel commented:
This is a WIP because I'd like to do the following before landing: 
* Land #5 and rebase this on top of that.
* Add an implementation for the dogstatsd module interface.(currently, this is only the interface)
   * I'm currently working on this implementation.

bradlangel committed:
Symlink to statsd_client_core from extended dir.

bradlangel committed:
Functorize IO module for dogstatsd implementation

bradlangel committed:
Implements dogstatsd metric modules

bradlangel committed:
Modify Dogstatsd metric set to be an integer

bradlangel committed:
Adds default for tags param in Metric.send

bradlangel committed:
Moves ServiceCheck.status type into its own module

bradlangel committed:
Implements dogstatsd.ml ServiceCheck modules

bradlangel committed:
Implements dogstatsd.ml Event modules

bradlangel committed:
Removes Base module from dogstatsd.ml

Base module was a wrapper module around a few modules, however, it
wasn't necessary as it didn't expose a type or do anything but be
syntactic candy. E.g. Base.Metric.t becomes Metric.t

bradlangel committed:
Moves dogstatsd.ml to statsd-client directory

bradlangel committed:
Adds Dogstatsd module to statsd-client-async lib

bradlangel committed:
Adds Dogstatsd module to statsd-client-lwt lib

bradlangel commented:
* Adds implementation for dogstatsd to both async and lwt modules
* This is based ontop of #3 and #5 as a result there are currently a few extra commits.

bradlangel committed:
Fixes comment in dogstatsd.mli

bradlangel committed:
Update maintainer and author information
